fw/http_parser.c

```diff
@@ -5926,68 +5917,78 @@ do {									\
 	__fsm_n = __try_str(&parser->hdr, chunk, p, __data_remain(p),	\
 			    str, sizeof(str) - 1);			\
 	if (__fsm_n > 0) {						\
 		if (chunk->len == sizeof(str) - 1) {			\
-			lambda1;					\
 			TRY_STR_INIT();					\
-			__FSM_H2_I_MOVE_BY_REF_NEQ_LAMBDA_n(state, __fsm_n, lambda2);	\
+			p += __fsm_n;					\
+			if (__data_off(p) < len)			\
+				goto *next_st;				\
+			if (likely(fin))				\
+				goto *next_st;				\
+			parser->_i_st = next_st;			\
+			__msg_hdr_chunk_fixup(data, len);		\
+			__FSM_I_chunk_flags(TFW_STR_HDR_VALUE);		\
+			__FSM_EXIT(CSTR_POSTPONE);			\
 		}							\
 		if (likely(fin))					\
 			return CSTR_NEQ;				\
+		parser->_i_st = curr_st;				\
 		__msg_hdr_chunk_fixup(data, len);			\
 		__FSM_I_chunk_flags(TFW_STR_HDR_VALUE);			\
-		finish;							\
 		__FSM_EXIT(CSTR_POSTPONE);				\
 	}
 
-#define H2_TRY_STR_BY_REF(str, curr_st, next_st)			\
-	H2_TRY_STR_LAMBDA_BY_REF_finish(str, { }, {			\
-		__FSM_EXIT(CSTR_NEQ);					\
-	}, {								\
-		parser->_i_st = curr_st;				\
-	}, next_st)
-
 /**
  * The same as @H2_TRY_STR_2LAMBDA(), but with explicit chunks control;
  * besides, @str must be of plain @TfwStr{} type and variable @fld is
  * used (instead of hard coded header field).
  */
-#define H2_TRY_STR_2LAMBDA_fixup(str, fld, lambda1, lambda2, curr_st, next_st) \
-	BUG_ON(!TFW_STR_PLAIN(str));					\
-	if (!chunk->data)						\
-		chunk->data = p;					\
-	__fsm_n = __try_str(fld, chunk, p, __data_remain(p),		\
-			    (str)->data, (str)->len);			\
-	if (__fsm_n > 0) {						\
-		if (likely(chunk->len == (str)->len)) {			\
-			lambda1;					\
-			TRY_STR_INIT();					\
-			__msg_field_fixup_pos(fld, p, __fsm_n);		\
-			__FSM_I_field_chunk_flags(fld, TFW_STR_HDR_VALUE); \
-			if (__data_off(p + __fsm_n) < len) {		\
-				p += __fsm_n;				\
-				goto next_st;				\
-			}						\
-			if (likely(fin))				\
-				lambda2;				\
-			parser->_i_st = &&next_st;			\
-			__FSM_EXIT(CSTR_POSTPONE);			\
-		}							\
-		if (likely(fin))					\
-			__FSM_EXIT(CSTR_NEQ);				\
-		parser->_i_st = &&curr_st;				\
-		__msg_field_fixup_pos(fld, p, __fsm_n);			\
-		__FSM_I_field_chunk_flags(fld, TFW_STR_HDR_VALUE);	\
-		__FSM_EXIT(CSTR_POSTPONE);				\
-	}
-
-#define H2_TRY_STR_LAMBDA_fixup(str, fld, lambda, curr_st, next_st)	\
-	H2_TRY_STR_2LAMBDA_fixup(str, fld, {}, lambda, curr_st, next_st)
+#define H2_TRY_STR_3LAMBDA_fixup(str, fld, lambda1, lambda2, lambda3,		\
+				 curr_st, next_st)				\
+	BUG_ON(!TFW_STR_PLAIN(str));						\
+	if (!chunk->data)							\
+		chunk->data = p;						\
+	__fsm_n = __try_str(fld, chunk, p, __data_remain(p),			\
+			    (str)->data, (str)->len);				\
+	if (__fsm_n > 0) {							\
+		if (likely(chunk->len == (str)->len)) {				\
+			lambda1;						\
+			TRY_STR_INIT();						\
+			__msg_field_fixup_pos(fld, p, __fsm_n);			\
+			__FSM_I_field_chunk_flags(fld, TFW_STR_HDR_VALUE);	\
+			if (__data_off(p + __fsm_n) < len) {			\
+				p += __fsm_n;					\
+				goto next_st;					\
+			}							\
+			parser->_i_st = &&next_st;				\
```

Comment:
Why did you move the line before `if (fin)`? Probably we need a comment here to note the importance of the code placement.

Reply:
I think that I was wrong here. I considered that there was a sufficient condition to change the state. But for this, the chunk must not be fin.